File: python/Greetings_wScrubbing/LogAux.py
def Scrub(s):
    return "".join([c if c.isascii() and c.isprintable() else '?' for c in s])

def Obscure(s):
    if len(s) < 5:
        return s[0] + '***' + s[len(s) - 1]
    return s[0] + '*' * (len(s) - 2) + s[len(s) - 1]

def ObscureUsername(s):
    parts = s.split('\\')
    return '\\'.join([Obscure(part) for part in s.split('\\')])

File: java/Greetings_wScrubbing/LogAux.java
public final class LogAux
{
	public static String Scrub(String a_str) {
		StringBuilder sb = new StringBuilder();
		for (int i = 0 ; i < a_str.length() ; ++i) {
			sb.append(Character.isISOControl(a_str.charAt(i)) ? '?' : a_str.charAt(i));
		}
		return sb.toString();
	}
	
	public static String Obscure(String a_str) {
		if (a_str.length() < 5) {
			return a_str.charAt(0) + "***" + a_str.charAt(a_str.length() - 1);
		}
		StringBuilder sb = new StringBuilder();
		sb.append(a_str.charAt(0));
		for (int i = 1 ; i < a_str.length() - 1 ; ++i) {
			sb.append('*');
		}
		sb.append(a_str.charAt(a_str.length() - 1));
		return sb.toString();
	}
	
	public static String ObscureUsername(String a_str) {
		String[] parts = a_str.split("\\\\");
		if (parts.length == 1) {
			return Obscure(a_str);
		}
		return Obscure(parts[0]) + "\\" + Obscure(parts[1]);
	}
}